File: src/ios/Agent/Backup/BackupJSONLWriter.swift
import Foundation

/// Append-only JSONL writer with automatic shard rollover (§2).
///
/// Every record is one self-contained line, so the importer can parse
/// line-by-line and skip a bad record without losing the file (§2.2 rule 3).
/// Writes go straight to a FileHandle — nothing accumulates in memory, which is
/// what lets a 100k-message export stay flat.
final class BackupJSONLWriter {
    private let directory: URL
    /// `messages` → `messages.jsonl`, then `messages-0002.jsonl` on rollover.
    private let baseName: String
    private let maxShardBytes: Int

    private var handle: FileHandle?
    private var currentURL: URL?
    private var currentBytes = 0
    private var shardIndex = 1

    private(set) var writtenRecords = 0
    private(set) var totalBytes: Int64 = 0
    private(set) var shardPaths: [String] = []

    private let encoder: JSONEncoder = {
        let e = JSONEncoder()
        // Deterministic key order keeps package diffs readable and makes the
        // integrity hash reproducible for the same input.
        e.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        e.dateEncodingStrategy = .iso8601
        return e
    }()

    init(directory: URL, baseName: String, maxShardBytes: Int = BackupFormat.maxShardBytes) {
        self.directory = directory
        self.baseName = baseName
        self.maxShardBytes = maxShardBytes
    }

    func write<T: Codable>(_ envelope: BackupRecordEnvelope<T>) throws {
        var line = try encoder.encode(envelope)
        line.append(0x0A)  // newline — the record separator
        try append(line)
        writtenRecords += 1
    }

    private func append(_ data: Data) throws {
        if handle == nil || currentBytes + data.count > maxShardBytes {
            try rollover()
        }
        guard let handle else { throw BackupError.writeFailed("no open shard for \(baseName)", underlying: nil) }
        try handle.write(contentsOf: data)
        currentBytes += data.count
        totalBytes += Int64(data.count)
    }

    private func rollover() throws {
        try close()
        let name = shardIndex == 1 ? "\(baseName).jsonl"
                                   : String(format: "%@-%04d.jsonl", baseName, shardIndex)
        let url = directory.appendingPathComponent(name)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        FileManager.default.createFile(atPath: url.path, contents: nil)
        handle = try FileHandle(forWritingTo: url)
        currentURL = url
        currentBytes = 0
        shardIndex += 1
        shardPaths.append(name)
    }

    func close() throws {
        try handle?.close()
        handle = nil
        currentURL = nil
    }

    /// True when nothing was ever written, so the caller can avoid emitting an
    /// empty file (and an empty category) into the package.
    var isEmpty: Bool { writtenRecords == 0 }
}

/// Small helper for the one-shot JSON files (`provider_config.json`,
/// `mcp_servers.json`, `manifest.json`).
enum BackupJSONFile {
    static func write<T: Encodable>(_ value: T, to url: URL) throws {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes, .prettyPrinted]
        encoder.dateEncodingStrategy = .iso8601
        let data = try encoder.encode(value)
        try FileManager.default.createDirectory(at: url.deletingLastPathComponent(),
                                                withIntermediateDirectories: true)
        try data.write(to: url, options: .atomic)
    }
}

File: src/android/app/src/main/java/com/openminis/app/backup/BackupJsonlWriter.kt
package com.openminis.app.backup

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import java.io.BufferedOutputStream
import java.io.File

/**
 * Append-only JSONL writer with automatic shard rollover (§2), mirroring
 * `src/ios/Agent/Backup/BackupJSONLWriter.swift`.
 *
 * Every record is one self-contained line, so the importer can parse
 * line-by-line and skip a bad record without losing the file (§2.2 rule 3).
 * Writes go straight to a stream — nothing accumulates in memory, which is
 * what lets a 100k-message export stay flat.
 *
 * @param baseName `messages` → `messages.jsonl`, then `messages-0002.jsonl`.
 */
class BackupJsonlWriter(
    private val directory: File,
    private val baseName: String,
    private val maxShardBytes: Int = BackupFormat.MAX_SHARD_BYTES,
) : AutoCloseable {

    private var stream: BufferedOutputStream? = null
    private var currentBytes = 0
    private var shardIndex = 1

    var writtenRecords = 0
        private set
    var totalBytes = 0L
        private set
    val shardPaths = mutableListOf<String>()

    /** True when nothing was ever written, so the caller can avoid emitting an empty category. */
    val isEmpty: Boolean get() = writtenRecords == 0

    /**
     * Write one record wrapped in the §2.2 rule-3 envelope: `t` dispatches on
     * the way in, `v` allows per-record migration.
     */
    fun write(type: String, version: Int = 1, payload: JsonElement) {
        val envelope = buildJsonObject {
            put("t", JsonPrimitive(type))
            put("v", JsonPrimitive(version))
            put("d", payload)
        }
        val line = (BackupFormat.json.encodeToString(JsonObject.serializer(), envelope) + "\n")
            .toByteArray(Charsets.UTF_8)
        append(line)
        writtenRecords += 1
    }

    private fun append(data: ByteArray) {
        if (stream == null || currentBytes + data.size > maxShardBytes) rollover()
        val out = stream ?: throw BackupException("no open shard for $baseName")
        out.write(data)
        currentBytes += data.size
        totalBytes += data.size
    }

    private fun rollover() {
        close()
        val name = if (shardIndex == 1) "$baseName.jsonl"
        else String.format("%s-%04d.jsonl", baseName, shardIndex)
        directory.mkdirs()
        stream = BufferedOutputStream(File(directory, name).outputStream())
        currentBytes = 0
        shardIndex += 1
        shardPaths.add(name)
    }

    override fun close() {
        stream?.flush()
        stream?.close()
        stream = null
    }
}
